### PR TITLE
feature/WIK 100 password reset backend

### LINKED ISSUE #638: [FEATURE] [BACKEND] Password reset [WIK-100]
### Describe the user story:
A front end user should be able to integrate password reset process with the backend apis

### Tasks (where applicable)

- [ ] Validate email and/or phone numbers for password reset request
- [ ] Set 1hr expiry time for a password reset token
- [ ] Send an email with a password reset token
- [ ] Send email notification for a successful password reset
- [ ] Update password api

### Tests (where applicable)

- [ ] Validate password reset token expiry time
- [ ] Email sending
- [ ] User password reset using token

```JS
const tokenSchema ={
  userId: {
    type: Schema.Types.ObjectId,
    required: true,
    ref: "user.id",
  },
  token: {
    type: String,
    required: true,
  },
  createdAt: {
    type: Date,
    default: Date.now,
    expires: 3600
  }
}
```

### LINKED ISSUE #639: [FEATURE] [BACKEND] Email Verification [WIK-98]
### Describe the user story:

On signup, a user should receive an email to verify their email

### Tasks (where applicable)

- [ ] On signup send email with verification token
- [ ] Update email verification status to ``tue``

### Tests (where applicable)

- [ ] Email verification token expiry time
- [ ] Send email on user creation
- [ ] Update valid email address

### Acceptance Criteria

- [ ] eg end-point  " api/users" should not list users  for non-admin users
- [ ] eg end-point  " api/users" should list users for admin users

### PR DESCRIPTION
**Pull request by mosesokemwa**
- endpoints for resetting passwords
- Co-authored-by: Brian Marete <brianmarete@users.noreply.github.com>

- [x] I have read [Wikonnects contributing guidelines](https://github.com/tunapanda/wikonnect/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update README.md`).
- [x] My pull request passes all tests.

Closes #638
Closes #639

### CODE REVIEW

**Review by brianmarete — APPROVED** [review]

File: server/routes/users.js [hunk 1]
```diff
@@ -424,30 +413,30 @@ router.post('/verify', async ctx => {
...
  *
  */
-router.get('/verify', async ctx => {
+
+router.get('/verify', permController.requireAuth, async ctx => {
   const decodedMail = Buffer.from(ctx.query.email, 'base64').toString('ascii');
-  const user = await User.query().findOne({ 'email': decodedMail, 'resetPasswordToken': ctx.query.token });
-  ctx.assert(user, 'No email found');
-  let message;
+  let user = await User.query().findOne({ 'email': decodedMail });
```
Comment (0xMurage) [inline server/routes/users.js]: someone can easily bypass verification by making an API request to this endpoint with email query parameter e.g. `/api/v1/users/${randomId}/verify?email=${base64EncodedEmail}`

the endpoint should verify that the token received is associated to the user email first.

File: server/routes/users.js [hunk 2]
```diff
@@ -424,30 +413,30 @@ router.post('/verify', async ctx => {
...
+  ctx.assert(user, 404, 'No email found');
+  let verifiedData;
   try {
     if (new Date() < user.resetPasswordExpires) {
```
Comment (0xMurage) [inline server/routes/users.js:460]: throw an error if the request has expired. With current flow, the endpoint will return 200 ok status

**Review by 0xMurage — CHANGES_REQUESTED** [review]

**Review by 0xMurage — APPROVED** [review]

**Comment by 0xMurage** [comment]
@brianmarete just noticed `resetPasswordToken` & `resetPasswordExpires`  properties are returned on `api/v1/users/` endpoints. I think we should not expose this to anyone{even to an admin}

**Comment by mosesokemwa** [comment]
> since this PR does away with Redis, this should also be updated on the readme.md.
> 
> Instructions on how to set up the mail configuration should also be included

This was already accounted for in the docs

**Comment by mosesokemwa** [comment]
> @brianmarete just noticed `resetPasswordToken` & `resetPasswordExpires` properties are returned on `api/v1/users/` endpoints. I think we should not expose this to anyone{even to an admin}

@brianmarete I added the attributes to secure field attributes in the model file, they will not be exposed

**Pull request merged**